src/allocation.rs

```diff
@@ -0,0 +1,134 @@
+//! Stuff to boost things in the `alloc` crate.
+
+use super::*;
+use alloc::{boxed::Box, rc::Rc, sync::Arc, vec::Vec};
+
+/// As [`try_cast_arc`](try_cast_arc), but unwraps for you.
+#[inline]
+pub fn cast_arc<A: Pod, B: Pod>(input: Arc<A>) -> Arc<B> {
+  try_cast_arc(input).map_err(|(e, _v)| e).unwrap()
+}
+
+/// Attempts to cast the content type of a [`Arc`](alloc::sync::Arc).
+///
+/// On failure you get back an error along with the starting `Arc`.
+///
+/// ## Failure
+///
+/// * The start and end content type of the `Arc` must have the exact same
+///   alignment.
+/// * The start and end size of the `Arc` must have the exact same size.
+#[inline]
+pub fn try_cast_arc<A: Pod, B: Pod>(input: Arc<A>) -> Result<Arc<B>, (PodCastError, Arc<A>)> {
+  if align_of::<A>() != align_of::<B>() {
+    Err((PodCastError::AlignmentMismatch, input))
+  } else if size_of::<A>() != size_of::<B>() {
+    Err((PodCastError::SizeMismatch, input))
+  } else {
+    // Note(Lokathor): This is much simpler than with the Vec casting!
+    let ptr: *mut B = Arc::into_raw(input) as *mut B;
+    Ok(unsafe { Arc::from_raw(ptr) })
+  }
+}
```

Comment:
This makes me all kinds of nervous.  Specifically if A is stored inside of something like this under the hood:
```rs
// NOTE: repr(rust)
struct ArcRefCounted<A> {
    rc: AtomicUsize,
    data: A,
}
```
The position of rc wouldn't be guaranateed to be compatible between A and B to my understanding.  If this pointer casting *is* sound/guaranteed by the compiler, I'd like to see some references in the comments!

Reply:
...the real type is ArcInner and it *is* repr(rust)... paranoia senses tingling...

Reply:
Yeah, you're right. and `Rc` has this same problem.

Reply:
This also happens with .as_mut() FWIW but glad I'm not missing anything